05_debug/debug_module.c

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/ctype.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/printk.h>
+
+#include <linux/sysfs.h>
+#include <linux/proc_fs.h>
+#include <linux/device.h>
+#include <linux/uaccess.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+MODULE_AUTHOR("Andrii Synenko");
+MODULE_DESCRIPTION("List data module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+
+#define PROCFS_BUFF_SIZE 255
+#define PROCFS_DIR_NAME	"debug_module"
+#define PROCFS_ENT_NAME	"upper"
+#define SYSFS_DIR_NAME	PROCFS_DIR_NAME
+#define SYSFS_ENT_NAME	"lower"
+#define STAT_ENT_NAME	"stat"
+
+struct convertor_struct {
+	char *str_buff;
+
+	int converted_last;
+	int converted_total;
+	int processed_last;
+	int processed_total;
+
+} sysfs_conv, procfs_conv;
+
+static int convert_case(bool to_lover, struct convertor_struct *conv,
+						ssize_t str_size)
+{
+	size_t indx = 0;
+
+	conv->converted_last = 0;
+
+	if (conv == NULL || conv->str_buff == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	while (conv->str_buff[indx] != '\0' &&
+		conv->str_buff[indx] != '\n' &&
+		indx < str_size) {
+		/*
+		 *!( isupper XOR to_lover )
+		 * |   t      |    t      | t - lovercase output
+		 * |   f      |    f      | t - uppercase output
+		 */
+		if (!(isupper(conv->str_buff[indx]) ^ to_lover)) {
```

Comment:
At first, `isupper()` returns **int** value which can be, say, `0x20`.  Common `ctype.h` functions' implementation contains 256-byte array with bit masks that describes characters and `is`-function just returns a value from the array masked with appropriate mask (`isalnum` uses an OR of UPPER, LOWER and DIGIT masks). You should not relay on actual UPPER mask even it is 0x01 in the current implementation.
Use an idiom`!!` to convert to bool.

But... As far as I remember, `toupper` functions guarantees that it return unchanged argument if it isn't a lowercase character, the same for `tolower`. This `if` is redundant. Perhups, this code will be more compact and fast. Perhups not :-)
```c
    char ch = to_lower ? tolower(...) : toupper(...);
    if (ch != conv->str_buff[indx])
        conv->converted_last++;
    conv->str_buff[indx] = ch;
```

Reply:
Changed to suggested solution